### ISSUE TITLE
Students can easily tell which problems they have finished and if any are left to complete from the problem set screen

### ISSUE THREAD

**Issue by benetechMatt** [benetech/MathShare#966]
Students can easily tell which problems they have finished and if any are left to complete from the problem set screen. problem is grayed out (but still clickable). Problems are marked finished when the student has added at least one step and selected the “finished” button at the end of the problem.

**Comment by benetechMatt** [comment]
see mockup in #967

**Comment by benetechMatt** [comment]
a green bar is placed at the bottom of the card along with the answer/equation of final step to indicate if a problem is finished

**Comment by alexrcabral** [comment]
If the green bar is displayed that should be enough that the greying out won't be necessary.

**Comment by alexrcabral** [comment]
See comments on mockup in #967

**Comment by benetechMatt** [comment]
questions resolved in #967 
Will keep the green bar and not gray out items.

**Comment by alexrcabral** [comment]
Good to go, assuming following the updates to the mockup in #967

**Comment by jscholes** [comment]
We are awaiting comments from @bpmcneilly for both this and #967.  But I can say upfront that if the green bar uses colour alone to indicate any sort of progress, it won't be accessible to a screen reader user so a textual alternative must also be provided.  In addition, care must be taken to ensure that the communication of progress is not inadvertently hidden when users enable a high contrast colour scheme.

**Comment by alexrcabral** [comment]
@jscholes 
The colour does not indicate progress nor change over time. Its purpose is to further distinguish between the top portion of the tile (the prompt) and the bottom portion with the solution. The only indication of progress is the existence or lack of a solution at the bottom of the tile, which is inherently textual. 

The only issue I could think of there is if it's difficult to differentiate auditorily between the first step of the problem and the solution if they're read out in one go, since both are math-based text. If there are three lines presented on a tile -- the instructions, the prompt, and the solution -- would a student using a screen reader be able to differentiate effectively between these lines?

**Comment by benetechMatt** [comment]
When being read aloud, would it be helpful to first state if the problem has been completed?
Here's an example of what could be read:

Problem 1 Completed
Find the value of x
3x+10=9
Final step is x=3

**Comment by jscholes** [comment]
I'd say it definitely needs to be indicated if it is visually obvious.  That could be via text directly indicating completion state, and/or headings to separate the prompt from the solution.  Many users won't be used to interacting with math content, so anything we can do to signpost meaning and context is welcome.

**Comment by benetechMatt** [comment]
@jscholes would the example I put in the previous step meet those needs?
@alexrcabral I'm thinking it would make sense to add TTS for the problem cards. One button to have the whole card read aloud and having it read similar to how it would for screen reader users as its going through multiple lines

**Comment by sinabahram** [comment]
How are you going to generate the speech for this? As in, how is the mathematics going to be encoded as text to send to the TTS, using SRE?

**Comment by jscholes** [comment]
This is the problem set view, so shares concerns raised and discussed in https://github.com/benetech/MathShare/issues/1025.

I also take issue with this:

> problem is grayed out (but still clickable).

This is non-standard behaviour.  If something is grayed out, people don't expect clicking it to do anything.  And people don't expect links to be grayed out full-stop, which is why an `<a>` element in HTML doesn't support the disabled attribute.

It also raises a question about how this can be surfaced for screen reader users, which links in with the discussions from https://github.com/benetech/MathShare/issues/1025.  If you do gray it out, you'll also have to programmatically indicate the disabled state for a screen reader user.  But as soon as you do that via `aria-disabled`, all interaction bets are off.

I suggest that the links for completed problems should link to the read-only problem view, have a textual indication of completed state for screen reader users, and be visually distinguished via something which doesn't involve making them look disabled.

**Comment by jscholes** [comment]
However, I do need to stress that I'm only addressing the surfacing of completion state here.  How users should interact with math within problem links is out of scope for this thread, but very much related and an open problem.

**Comment by benetechMatt** [comment]
@jscholes will not be doing the grayed out text/box

@alexrcabral adding all of the proposed copy for all users seems overwhelming. Not having it would be an issue for TTS synchronized highlighting.  Suggestions?

**Comment by jscholes** [comment]
@benetechMatt I do wonder if we're trying to cram too much functionality into a single screen.  This is a listing view, but we're already saying that it must have:

1. Math that can be accessibly explored by a screen reader user even though it's in a link;
2. TTS functionality;
3. the solution to a problem if it's been completed;
4. an accessible indication of completion; and
5. an accessible reordering mechanism (#764).

I think 4 and 5 definitely fit.  I think 1 could be avoided, if every problem had a unique textual title (which may not be possible).  But I think 2 and 3 will make the UI far too busy.  If students want to explore their solution or have something read out, would it be problematic to just have them click on the problem to access those functions?

You could then have text and iconography, in whatever configuration works, to indicate completion.

**Comment by jscholes** [comment]
If all of this functionality does make into the problem set listing, we can make it accessible.  But the current simple unordered list of links may have to be exchanged for a different pattern.  For example, a heading for each problem title which is the link, a reviewable MathML area underneath, etc.

**Comment by benetechMatt** [comment]
For this issue, we won't enable built in TTS. 
None screen reader users can tell which problems are completed by seeing the final step equation in at the bottom of the link with a green background.
Screen reader users have the first line read out as "Problem x completed", then the question, equation, and final step equation

Mockup as shown in https://github.com/benetech/MathShare/issues/967#issuecomment-601893505

**Comment by bpmcneilly** [comment]
Based on the amount of information we're talking about in this view, I think I'm in line with @jscholes about adding a separate pattern for organizing the problem set listings.

While TTS isn't being required, there is still a lot of content being conveyed in each card, enough that I think we may need to rethink the unordered list. Using problem titles as headings and links to begin each card could then allow us to remove the interaction from mathML content (problems or solutions), and provide a clearer indication of completion status / potentially the truncation expanding controls.

**Comment by bpmcneilly** [comment]
Also, to follow up on @benetechMatt's previous comment, would the 'Problem x completed' be displayed visually, or off-screen? If we're taking an approach where we're adding more structure to these cards, this message could be added visually above the solution mathML to help cement the purpose of the content on the screen.

While the visual spacing between initial problem & solution helps to distinguish the two, if users of high contrast modes aren't exposed to the background green color, it may not be as intuitive to determining what this content is the first time it's seen.

**Comment by jscholes** [comment]
Agreed with @bpmcneilly.  In addition, @benetechMatt if the only thing we remove is the TTS button, it feels like we haven't significantly made the view any easier to digest, while irritating any user who needs that button to read the math.  They'll now need to activate the problem link, even though everyone else doesn't need to.

So, in summary, either:

* keep the unordered list, but significantly cut down the information being displayed; or
* devise a more structured layout, making sure that everything on show is accessible and equitable.

**Comment by benetechMatt** [comment]
how the problems should be read by a screen reader.

Problem 1 complete. Find the value of x.
First step. 3x+10=9.
Answer. x=3.

what is the right level of pause between the three sections of the math tile and parts of a line such as "Answer" and "x=3". How should we denote it when scoping out the issue and implementation?

**Comment by sinabahram** [comment]
Hi, I’m really confused by this question. how are you planning on affecting the level of pause here? You don’t have control over how long a screen reader pauses.

 

Also, isn’t the answer x = -1/3? Not that it matters, but that was bothering me 😊.

**Comment by jscholes** [comment]
@benetechMatt and @alexrcabral - We've provided all accessibility comments and questions from our side.  It would be good to discuss this further, on a call if necessary, because it and the related tickets are relatively complex.  Any thoughts?
